sdk/storage/faq.md

```diff
@@ -0,0 +1,374 @@
+then upload the chunks with multiple threads.
+This one suits in most cases. You can expect higher throughput because the chunks are transferred concurrently. It's especially recommended if you need to tansfer large blobs efficiently.
+
+## How to efficiently upload large amount of small blobs?
+
+Unfortunately this SDK doesn't provide a convenient way to upload many blobs or directory contents (files and sub-directories) with just one function call.
+You have to create multiple threads, traverse the directories by yourself and upload blobs one by one in each thread to speed up the transfer.
+Or you can use tools like [AzCopy](https://learn.microsoft.com/en-us/azure/storage/common/storage-ref-azcopy) or Data Movement Library.
+
+## How to ensure data integrity with transactional checksum?
+
+Generally speaking, TLS protocol includes checksum that's strong enough to detect accidental corruption or deliberate tampering.
+Another layer of checksum is usually not necessary if you're using HTTPS.
+
+If you really want it for whatever reason, for example, to detect corruptions before the data is written to the socket,
+you can leverage transactional checksum feature in the SDK.
+With this feature, you provide a pre-calculated MD5 or CRC64 checksum when calling an upload API,
+storage service will calculate checksum after it receives the data and compare with the one you provide,
+and fail the request if they don't match.
+Make sure you calculate the checksum early enough to cover the time when the corruption may happen.
+
+This functionality also works for download operations.
+Below is a code sample to use this feature.
+
+```C++
+// upload data with pre-calculated checksum
+Blobs::UploadBlockBlobOptions options;
+options.TransactionalContentHash = ContentHash();
+options.TransactionalContentHash.Value().Algorithm = HashAlgorithm::Crc64;
+options.TransactionalContentHash.Value().Value = crc64;  // CRC64 checksum of the data in Base64 encoding
+blobClient.Upload(stream, options);
+
+// download and verify checksum
+Blobs::DownloadBlobOptions options;
+options.RangeHashAlgorithm = HashAlgorithm::Crc64;
+auto response = blobClient.Download(options);
+auto crc64 = response.Value.TransactionalContentHash.Value();
+// Now you can verify checksum of the downloaded data
+```
+
+## Thread safety guarantees
+
+All storage client APIs are thread-safe.
+You can call the same API or different APIs on the same client instance from multiple threads without additional synchronization.
+It's guaranteed to either return a successful response or throw an exception.
+
+However, this doesn't mean access to underlying storage service has no race condition.
+It is still possible that you'll get a 409 error if you read a file that's being written by another party. (For example, one thread reading an append blob while another thread appending data to it.)
+
+## Http pipeline and policies
```

Comment:
Should we add a warning to prevent users from modifying the header directly in the HttpTransport which may cause StringToSign failure?